Add metadata fields for mappings (content gap initiative)

vagimeli opened the pull request:
### Description
Add documentation to address content gaps

### Issues Resolved
Closes #4300 
See also PR #7115 

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vagimeli commented:
Status = Need SME point of contact for tech review and feedback on the following questions: 

1. Need SME input about mapping limits settings, their default values, and how to configure them if needed.
2. Need SME input about removal of mapping types: Does OpenSearch allow for defining mapping types within an index? If so, SME needs to provide information to include in documentation. Do we have guidance on how to structure data without relying on mapping types?

mgodwan commented on _field-types/index.md:
```diff
@@ -88,6 +82,41 @@ POST sample-index1/_mapping
+
+Parameter | Description
+:--- | :---
+`analyzer` | Specifies the analyzer used to analyze string fields.
+`boost` | Specifies a field-level query time to boost.
+`coerce` | Tries to convert the value to the specified data type.
+`copy_to` | Copies the values of this field to another field.
+`doc_values` | Specifies whether the field should be stored on disk to make sorting and aggregation faster.
+`dynamic` | Determines whether new fields should be added dynamically.
+`enabled` | Specifies whether the field is enabled or disabled. 
+`format` | Specifies the date format for date fields. 
+`ignore_above` | Skips indexing values that are longer than the specified length.
+`ignore_malformed` | Specifies whether malformed values should be ignored. 
+`index` | Specifies whether the field should be indexed.
+`index_options` | Specifies what information should be stored in the index for scoring purposes.
```
do you think it will be helpful to add possible values for each of these parameters (or examples in case the value set is not limited)

vagimeli replied:
@mgodwan Yes to both.

vagimeli replied:
What are the default values?

mgodwan replied:
The default values for each of these may vary across different field types (e.g. keyword, text may have different defaults for different parameters)

For each field type documentation (e.g. text, integer, geopoint), we can then share what the default/allowed values for the parameters are.

vagimeli replied:
@mgodwan Please verify the default and allowed values are accurate and make any necessary changes, as needed. Thank you.

mgodwan commented on _field-types/index.md:
```diff
@@ -88,6 +82,41 @@ POST sample-index1/_mapping
+`index` | Specifies whether the field should be indexed.
+`index_options` | Specifies what information should be stored in the index for scoring purposes.
+
+## Mapping limit settings
+
+OpenSearch may have certain limits or settings related to mappings, such as the maximum number of fields allowed in an index or the maximum depth of nested objects. These settings can be configured based on your requirements. <_Need SME input about mapping limits settings, their default values, and how to configure them if needed._>
+
+## Removal of mapping types
+
+<_Need SME input. Does OpenSearch allow for defining mapping types within an index? If so, SME needs to provide information to include in documentation. Do we have guidance on how to structure data without relying on mapping types?_>
+
+## Runtime fields
+
+You can define fields at query time, rather than at index time, by using runtime fields. this can be useful for creating fields based on the values of other fields, or for performing transformations on data during the query process. Runtime fields are defined in the query itself and do not affect the underlying data in the index.
```
Ack. I can help you with the possible settings and examples (if applicable).

mgodwan replied:
Mapping Limit Settings
 [TBU Description column]

| Settings | Default Value | Allowed Values | Type | Description |
|-|-|-|-|-|
| index.mapping.nested_fields.limit | 50 | [0,) | Dynamic | |
| index.mapping.nested_objects.limit | 10000 | [0,) | Dynamic | |
| index.mapping.total_fields.limit | 1000 | [0,) | Dynamic | |
| index.mapping.depth.limit | 20 | [1,100] | Dynamic | |
| index.mapping.field_name_length.limit | 50000 | [1,50000] | Dynamic | |
| index.mapper.dynamic | true | {true,false} | Dynamic | |

vagimeli replied:
@mgodwan  I added descriptions. Please confirm the descriptions are accurate.

mgodwan commented:
> Does OpenSearch allow removal of mapping types and defining mapping types within an index?

Do you mean custom types here which were present in earlier versions on elasticsearch? If yes, the mapping types have been deprecated since OS 1.0 itself

vagimeli commented:
@mgodwan I've added first-pass drafts of the mappings documentation. When you review it, please confirm that the code examples are relevant (I tested them using Dev Tools, but am not sure if these examples are most relevant to our users.) Also, please verify that the narrative text is accurate and I haven't miscommunicated any information. Thank you!

mgodwan commented on _field-types/index.md:
```diff
@@ -88,6 +82,42 @@ POST sample-index1/_mapping
+| index.mapping.field_name_length.limit | 50000 | [1,50000] | Dynamic | Limits the maximum length of field names that can be defined in an index mapping. |
+| index.mapper.dynamic | true | {true,false} | Dynamic | Determines whether new fields should be added dynamically to the mapping when they are encountered in a document. |
+
+## Runtime fields
+
+You can define fields at query time, rather than at index time, by using runtime fields. this can be useful for creating fields based on the values of other fields, or for performing transformations on data during the query process. Runtime fields are defined in the query itself and do not affect the underlying data in the index.
+
```
This is not yet available on OpenSearch
https://github.com/opensearch-project/documentation-website/issues/6943

vagimeli replied:
Removed

mgodwan commented on _field-types/index.md:
````diff
@@ -220,3 +250,27 @@ The response contains the specified fields:
 }
 ```
+
+## Delete a mapping
````
I don't think we support `DELETE` verb on mappings.

mgodwan commented on _field-types/metadata-fields/id.md:
```diff
@@ -0,0 +1,84 @@
+
+## Querying on the `_id` field
+
+While the `_id` field is accessible in various queries, it is restricted from use in aggregations, sorting, and scripting. If you need to sort or aggregate on the `_id` field, it is recommended to duplicate the content of the `_id` field into another field that has `doc_values` enabled.
```
@sandeshkr419  Would you be able to confirm this from search perspective?

sandeshkr419 replied:
Hey @mgodwan missed the notification for this one somehow. 

The explanation seems correct. 

@vagimeli I'm wondering if we should link the usage example of querying over `_id` field as well: https://opensearch.org/docs/2.1/opensearch/query-dsl/term/#ids 


Something like:

> While the `_id` field is accessible in various queries (see [usage](https://opensearch.org/docs/2.1/opensearch/query-dsl/term/#ids)), it is restricted from use in aggregations, sorting, and scripting. If you need to sort or aggregate on the `_id` field, it is recommended to duplicate the content of the `_id` field into another field that has `doc_values` enabled.

vagimeli replied:
Revised.

mgodwan commented on _field-types/metadata-fields/index-metadata.md:
```diff
@@ -0,0 +1,74 @@
+  "name": "Widget X"
+}
+
+PUT customers/_doc/2?refresh=true
```
We can remove `refresh=true` param?

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+
+The `field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query remains functional but does not rely on `field_names`.
+
+## Mapping example
```
@mgodwan Please confirm this is a relevant example for OpenSearch users.

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+
+# Field names
+
+The `field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
```
@mgodwan Please review this narrative for technical accuracy.

vagimeli replied:
Revised.

vagimeli commented on _field-types/index.md:
```diff
@@ -14,33 +14,27 @@ redirect_from:
 :--- | :---
 null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
-boolean | OpenSearch accepts `true` and `false` as boolean values. An empty string is equal to `false.`
+boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```
```suggestion
Boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

vagimeli commented on _field-types/metadata-fields/id.md:
```diff
@@ -0,0 +1,84 @@
+
+## Querying on the `_id` field
+
+While the `_id` field is accessible in various queries, it is restricted from use in aggregations, sorting, and scripting. If you need to sort or aggregate on the `_id` field, it is recommended to duplicate the content of the `_id` field into another field that has `doc_values` enabled.
```
```suggestion
While the `_id` field is accessible in various queries, it is restricted from use in aggregations, sorting, and scripting. See [IDs query]({{site.url}}{{site.baseurl}}/query-dsl/term/ids/) for an example of the field's usage. If you need to sort or aggregate on the `_id` field, it is recommended to duplicate the content of the `_id` field into another field that has `doc_values` enabled.
```

mgodwan commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 integer | A signed 32-bit number.
 object | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.
-array | Arrays in OpenSearch can only store values of one type, such as an array of just integers or strings. Empty arrays are treated as though they are fields with no values.
+array | Arrays in OpenSearch can only store values of one type, such as an array of only integers or strings. Empty arrays are treated as though they are fields with no values.
```
We should highlight that there is no specific array type, and a set of values can be passed for fields

vagimeli replied:
Revised

mgodwan commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 object | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.
-array | Arrays in OpenSearch can only store values of one type, such as an array of just integers or strings. Empty arrays are treated as though they are fields with no values.
+array | Arrays in OpenSearch can only store values of one type, such as an array of only integers or strings. Empty arrays are treated as though they are fields with no values.
 text | A string sequence of characters that represent full-text values.
 keyword | A string sequence of structured characters, such as an email address or ZIP code.
 date detection string | Enabled by default, if new string fields match a date's format, then the string is processed as a `date` field. For example, `date: "2012/03/11"` is processed as a date.
 numeric detection string | If disabled, OpenSearch may automatically process numeric values as strings when they should be processed as numbers. When enabled, OpenSearch can process strings into `long`, `integer`, `short`, `byte`, `double`, `float`, `half_float`, `scaled_float`, and `unsigned_long`. Default is disabled.
 
```
Can we cover all field types from here? https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesModule.java#L151-L174

vagimeli replied:
Should we link the reader to the GitHub file instead of listing the mapping types in the documentation?

vagimeli replied:
Should we link to the GitHub link instead of listing the field types in the documentation?

mgodwan replied:
Linking to code can be tricky as the code can move due to refactoring making the documentation point to old/unusable links. Hence, I'd prefer to keep it within the documentation if possible.

mgodwan commented on _field-types/index.md:
````diff
@@ -14,38 +14,57 @@ redirect_from:
+
+The syntax for defining a dynamic mapping template in OpenSearch looks like the following:
+
+```json
+  "dynamic_templates": [
+    {
+      "template_name": { 
+        ...  match conditions ... 
+        "mapping": { ... } 
+      }
+    },
+    ...
+  ]
+```
````
I can share an actual example here. Let me know if that will be better

vagimeli replied:
Yes, please share an example.

mgodwan replied:
```
PUT index
{
  "mappings": {
    "dynamic_templates": [
        {
          "fields": {
            "mapping": {
              "type": "short"
            },
            "match_mapping_type": "string",
            "path_match": "status*"
          }
        }
    ]
  }
}
```

The above will create a field like `status_code` (based on path_match parameter) with data type `short` if during the first pass, it detects the input value to be a `string`

mgodwan reviewed (changes requested):
I'm still looking through the files. Have added some initial comments.

rursprung commented:
coming from #4300 i had hoped to find some more content on the `analyzer`, `normalizer`, etc. attributes in this PR but don't see them here. i think it'd be great if this could be documented a bit more as it seems that not all of these can be used with all field types (e.g. `normalizer` seems to work with `keyword` but not `text` fields while `analyzer` works with `text` but not `keyword`?). even just having the information that these are type-specific would be nice, and then additionally the per-type documentation should go into details. or maybe the overview could also contain a list of field types supporting each of the attributes?
these pseudo-generic attributes (exist everywhere but can't be set everywhere) are quite hard to deal with without (good) documentation

mgodwan commented:
@rursprung I see this PR as more of metadata fields being documented and not mapping params for field types.

See this PR for for mapping parameters: https://github.com/opensearch-project/documentation-website/pull/7115 which contain params like analyzer, doc_values, etc.

Once we add more documentation for each field type (e.g. keyword, text), we can mention there what params each field type supports.

navneet1v commented on _field-types/metadata-fields/source.md:
```diff
@@ -0,0 +1,55 @@
+
+## Including or excluding fields
+
+You can prune contents of the `_source` field after the document has been indexed but before the field is stored. You can use the `includes` and excludes` parameters to control which field are included in the stored `_source` field, as shown in the following example:
+
```
Disabling the single field also has same behavior as that of disabling the whole source. We should mention that in this section too.

rursprung commented on _field-types/metadata-fields/source.md:
```diff
@@ -0,0 +1,55 @@
+
+## Including or excluding fields
+
+You can prune contents of the `_source` field after the document has been indexed but before the field is stored. You can use the `includes` and excludes` parameters to control which field are included in the stored `_source` field, as shown in the following example:
```
```suggestion
You can prune contents of the `_source` field after the document has been indexed but before the field is stored. You can use the `includes` and `excludes` parameters to control which field are included in the stored `_source` field, as shown in the following example:
```

(randomly stumbled over this 😅 - noticed it because of the syntax highlighting on GH)

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 You can define how documents and their fields are stored and indexed by creating a _mapping_. The mapping specifies the list of fields for a document. Every field in the document has a _field type_, which defines the type of data the field contains. For example, you may want to specify that the `year` field should be of type `date`. To learn more, see [Supported field types]({{site.url}}{{site.ba ...
 
-If you're just starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.
+If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.
```
```suggestion
If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, it's important to note that dynamic mappings can have performance implications, especially for large datasets or high-volume indexing operations.

While dynamic mappings provide convenience and flexibility, it's generally recommended to use explicit mappings whenever possible. With explicit mappings, you define the exact structure and data types of your fields upfront, enforcing a consistent standard across your data. This approach not only ensures data integrity but also optimizes performance by allowing OpenSearch to efficiently store and query your data.

If you know exactly what types your data falls under and want to enforce that standard, then you should use explicit mappings. Even if you're unsure about your data structure initially, it is advisable to switch to explicit mappings once you have a better understanding of your data to avoid potential performance issues and maintain data consistency.
```

mgodwan replied:
LGTM

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 :--- | :---
 null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
-boolean | OpenSearch accepts `true` and `false` as boolean values. An empty string is equal to `false.`
+Boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```
```suggestion
`boolean` | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 
 Type | Description
 :--- | :---
 null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```
```suggestion
`null` | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 :--- | :---
-null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
-boolean | OpenSearch accepts `true` and `false` as boolean values. An empty string is equal to `false.`
+`null` | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```
```suggestion
null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 :--- | :---
 null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
-boolean | OpenSearch accepts `true` and `false` as boolean values. An empty string is equal to `false.`
+`boolean` | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```
```suggestion
boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 integer | A signed 32-bit number.
 object | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.
-array | Arrays in OpenSearch can only store values of one type, such as an array of just integers or strings. Empty arrays are treated as though they are fields with no values.
+array | Arrays in OpenSearch can only store values of one type, such as an array of only integers or strings. Empty arrays are treated as though they are fields with no values.
```
```suggestion
array | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements, not as fields with no values. OpenSearch supports querying and filtering arrays, including checking for values, range queries, and array operations like concatenation and intersection. Nested arrays, containing complex objects or other arrays, are also supported for advanced data modeling. 
```

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
 You can define how documents and their fields are stored and indexed by creating a _mapping_. The mapping specifies the list of fields for a document. Every field in the document has a _field type_, which defines the type of data the field contains. For example, you may want to specify that the `year` field should be of type `date`. To learn more, see [Supported field types]({{site.url}}{{site.ba ...
 
-If you're just starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.
+If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.
```
Revised

vagimeli commented on _field-types/index.md:
```diff
@@ -84,11 +105,30 @@ POST sample-index1/_mapping
 
+## Mapping parameters
+
+Mapping parameters are used to configure the behavior of fields in an index. See the [Mapping parameters](inert-link-to-page) page for more information.
```
```suggestion
Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}//field-types/) page for more information.
```

vagimeli commented on _field-types/index.md:
```diff
@@ -84,11 +105,30 @@ POST sample-index1/_mapping
 
+## Mapping parameters
+
+Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}//field-types/) page for more information.
```
```suggestion
Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}/field-types/) page for more information.
```

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+
+# Field names
+
+The `field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
```
```suggestion
The `_field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
```

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+---
+layout: default
+title: Field names
```
```suggestion
title: _field_names
```

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+
+The `field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
+
+The `field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query remains functional but does not rely on `field_names`.
```
```suggestion
`_field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query remains functional but does not rely on `field_names`.
```

vagimeli replied:
@mgodwan @rursprung Is the text at line 13 technically accurate? If not, please make suggested edit.

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+
+## Mapping example
+
+{
```
```suggestion
"mappings": { "_field_names": { "enabled": "true" }, "properties": {
```

vagimeli commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,84 @@
+  }
+  ```
+  {% include copy-curl.html %}
+
````
```suggestion
A unique `_id` is automatically generated and assigned if you do not provide one.
{: .note}
```

vagimeli commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,84 @@
+  }
+  ```
+  {% include copy-curl.html %}
+
````
Revised

vagimeli commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,84 @@
+    ]
+  }
+  ```
+  {% include copy-curl.html %}
````
```suggestion
  {% include copy-curl.html %}
  
```

mgodwan commented on _field-types/index.md:
```diff
@@ -14,38 +14,57 @@ redirect_from:
+Note the following: 
+
+- The template name can be any string value. 
+- The match conditions can include any of the following: `match_mapping_type`, `match`, `match_pattern`, `unmatch`, `path_match`, or `path_unmatch`. 
```
The supported params for matching conditions with `dynamic_templates` are:

`match_mapping_type`: Detected JSON type based on which mapping should be applied. (e.g. string, long, double, object, binary, boolean, date)
`match`: Regex for matching field name and apply mapping
`unmatch`: Regex to verify and not apply mapping if the field name satisfies the provided regex
`match_pattern`: Allowed values are `regex` and `simple`. Based on this, it applies the regex matching. Default is `simple`
`path_match`: Allows for nested path matching along with regex
`path_unmatch`: Allows for nested path matching along with regex
`mapping`: Mapping to be applied

vagimeli replied:
Added.

vagimeli commented on _field-types/index.md:
```diff
@@ -14,38 +14,61 @@ redirect_from:
+- The template name can be any string value. 
+- The match conditions can include any of the following: `match_mapping_type`, `match`, `match_pattern`, `unmatch`, `path_match`, or `path_unmatch`. 
+- Specify the `mapping` to be used for the matched field.
+
```
@mgodwan What are the defaults for these parameters?

mgodwan replied:
Defaults are null. These are mostly logical wrapper which provide some kind of filters (based on the description). If they are not provided, the filter is not used for determining if dynamic template should be used for a field or not.

vagimeli commented on _field-types/index.md:
````diff
@@ -110,6 +150,7 @@ PUT /test-index
 ```
+{% include copy-curl.html %}
 
 You can add a document that has a malformed IP address to your index:
````
Yes, we can move them.

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,42 @@
+
+# Field names
+
+The `field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
```
```suggestion
The `_field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
```

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,42 @@
+
+The `_field_names` field indexes the names of fields within a document that contain non-null values. This field support the `exists` query, which identifies documents with or without non-null values for a specified field. 
+
+`_field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query remains functional but does not rely on `field_names`.
```
```suggestion
`_field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query remains functional but does not rely on `_field_names`.
```

vagimeli commented on _field-types/metadata-fields/index.md:
```diff
@@ -0,0 +1,21 @@
+
+Metadata fields | Description
+:--- | :---
+`field_names` | The fields within the document that hold non-empty or non-null values.   
```
```suggestion
`_field_names` | The fields within the document that hold non-empty or non-null values.   
```

vagimeli commented:
@mgodwan I've addressed your feedback. Will you give this a final review and approval? Once you're good with it, I'll move it forward to editorial review and then publish it. Thank you for all your support :)

mgodwan commented:
Thanks for helping with improving the feature documentation :slightly_smiling_face:

vagimeli commented on _field-types/mappings-use-cases.md:
````diff
@@ -0,0 +1,123 @@
+```
+{% include copy-curl.html %}
+
+This indexed IP address does not throw an error because `ignore_malformed` is set to true. 
````
```suggestion
This indexed IP address does not throw an error because `ignore_malformed` is set to `true`. 
```

Naarcha-AWS commented on _field-types/index.md:
```diff
@@ -12,43 +12,77 @@ redirect_from:
+While dynamic mappings automatically add new data and fields, using explicit mappings is recommended. Explicit mappings let you define the exact structure and data types upfront. This helps maintain data consistency and optimize performance, especially for large datasets or high-volume indexing operations.
 
-For example, if you want to indicate that `year` should be of type `text` instead of an `integer`, and `age` should be an `integer`, you can do so with explicit mappings. By using dynamic mapping, OpenSearch might interpret both `year` and `age` as integers.
+Even if you are unsure about your data structure at first, switching to explicit mappings later is recommended once you understand your data better. This helps avoid potential performance problems and keeps your data consistent. For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mappi ...
```
This paragraph feels slightly redundant with "This helps maintain data consistency and optimize performance, especially for large datasets or high-volume indexing operations." from the previous paragraph. Maybe combine "For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mapping," with the previous paragraph and delete the rest?

```suggestion
For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mapping.
```

Naarcha-AWS commented on _field-types/index.md:
```diff
@@ -12,43 +12,77 @@ redirect_from:
+
+### Dynamic mapping parameters
+
+The `dynamic_templates` support the following parameters for matching conditions and mapping rules. The default vaule is `null`.
```
```suggestion
The `dynamic_templates` support the following parameters for matching conditions and mapping rules. The default value is `null`.
```

Naarcha-AWS commented on _field-types/index.md:
````diff
@@ -84,84 +120,29 @@ POST sample-index1/_mapping
-```
+| Setting | Default value | Allowed value | Type | Description |
+|-|-|-|-|-|
+| `index.mapping.nested_fields.limit` | 50 | [0,) | Dynamic | Limits the maximum number of nested fields that can be defined in an index mapping. |
````
Does `[0,)` mean that any value above `0` is allowed?

vagimeli replied:
@mgodwan Please see question above.

Naarcha-AWS commented on _field-types/index.md:
```diff
@@ -220,3 +203,8 @@ The response contains the specified fields:
+
+## Mappings use cases
+
+See use case examples, including mapping string fields and ignoring malformed IP addresses, in [Mappings use cases](). 
```
```suggestion
See use case examples, including mapping string fields and ignoring malformed IP addresses, in [Mappings use cases]({{site.url}}{{site.baseurl}}/field-types/mappings-use-cases/). 
```

Naarcha-AWS commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,42 @@
+
+## Mapping example
+
+{
```
This code example needs ```

Naarcha-AWS commented on _field-types/metadata-fields/ignored.md:
```diff
@@ -0,0 +1,148 @@
+
+The `_ignored` field helps you manage and understand issues related to malformed data in your documents. This field is used to index and store the name of fields within a document that were ignored during the indexing process, when the `ignore_malformed` setting is enabled in the [index mapping]({{site.url}}{{site.baseurl}}/field-types/#mapping-example-usage). 
+
+The `_ignored` field allows you to search and identify documents that contain fields that were ignored, as well as the specific field names that were ignored. The can be useful for troubleshooting and understadning issues related to malformed data in your documents. 
```
```suggestion
The `_ignored` field allows you to search and identify documents that contain fields that were ignored, as well as the specific field names that were ignored. The can be useful for troubleshooting and understanding issues related to malformed data in your documents. 
```

Naarcha-AWS commented on _field-types/metadata-fields/ignored.md:
```diff
@@ -0,0 +1,148 @@
+
+The `_ignored` field allows you to search and identify documents that contain fields that were ignored, as well as the specific field names that were ignored. The can be useful for troubleshooting and understadning issues related to malformed data in your documents. 
+
+You can query the `_ignored` field using `term`, `terms`, and `exists` queries, and the results will be included in the search hits.
```
```suggestion
You can query the `_ignored` field using the `term`, `terms`, and `exists` queries, and the results will be included in the search hits.
```

Naarcha-AWS commented on _field-types/metadata-fields/index-metadata.md:
```diff
@@ -0,0 +1,87 @@
+
+## Querying on the `_index` field
+
+The `_index` field is a special field that represents the index a document was indexed into. You can use this field in your queries to filter, aggregate, sort, or retrieve index information for your search results.
```
```suggestion
The `_index` field represents the index a document was indexed into. You can use this field in your queries to filter, aggregate, sort, or retrieve index information for your search results.
```

Naarcha-AWS commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,93 @@
+layout: default
+title: Meta
+nav_order: 30
+has_children: false
```
This piece of metadata isn't needed, since the parent is already specified and this page contains no grandparents.

```suggestion
```

Naarcha-AWS commented on _field-types/metadata-fields/index-metadata.md:
```diff
@@ -0,0 +1,87 @@
+layout: default
+title: Index
+nav_order: 25
+has_children: false
```
```suggestion

```

Naarcha-AWS commented on _field-types/metadata-fields/ignored.md:
```diff
@@ -0,0 +1,148 @@
+layout: default
+title: Ignored
+nav_order: 15
+has_children: false
```
```suggestion

```

Naarcha-AWS commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,87 @@
+      }
+    ]
+  }
+  ```
````
Moving the spacing to the left might fix the small formatting issue underneath.

Naarcha-AWS commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,42 @@
+layout: default
+title: _field_names
+nav_order: 10
+has_children: false
```
```suggestion

```

Naarcha-AWS commented on _field-types/metadata-fields/id.md:
```diff
@@ -0,0 +1,87 @@
+layout: default
+title: ID
+nav_order: 20
+has_children: false
```
```suggestion

```

Naarcha-AWS commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,96 @@
+layout: default
+title: Routing
+nav_order: 35
+has_children: false
```
```suggestion

```

Naarcha-AWS commented on _field-types/metadata-fields/source.md:
```diff
@@ -0,0 +1,55 @@
+layout: default
+title: Source
+nav_order: 40
+has_children: false
```
```suggestion

```

vagimeli commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,96 @@
+
+# Routing
+
+Documents are routed to specific shards within an index using a hashing algorithm. By default, the document's `_id` field is used as the routing value, but you can also specify a custom routing value for each document.
```
```suggestion
OpenSearch uses a hashing algorithm to route documents to specific shards within an index. By default, the document's `_id` field is used as the routing value, but you can also specify a custom routing value for each document.
```

Naarcha-AWS reviewed (approved):
A few small fixes and suggestions, otherwise LGTM.

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,41 @@
+      }
+    }
+  }
+}
```
```suggestion
}
```
```

vagimeli commented:
> A few small fixes and suggestions, otherwise LGTM.

@Naarcha-AWS Thank you for reviewing this PR. I accepted your suggestions and revised areas where you left comments.

vagimeli commented:
@natebower This PR is ready for editorial at your availability. Thank you.

Naarcha-AWS commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,86 @@
+```
+{% include copy-curl.html %}
+
+This returns returns both documents with `_id` values of `1` and `2`.
````
```suggestion
This returns both documents with `_id` values of `1` and `2`.
```

Naarcha-AWS commented on _field-types/metadata-fields/index.md:
```diff
@@ -0,0 +1,21 @@
+--
```
```suggestion
---
```

natebower commented on _field-types/metadata-fields/meta.md:
````diff
@@ -0,0 +1,87 @@
+```
+{% include copy-curl.html %}
+
+## Retrieving meta information
````
Should "meta" be in code font?

natebower commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,87 @@
+
+## Retrieving meta information
+
+You can retrieve the `_meta` information for an index using the [get mapping API]({{site.url}}{{site.baseurl}}/field-types/#get-a-mapping) operation, for example, as shown in the following request:
```
```suggestion
You can retrieve the `_meta` information for an index using the [get mapping API]({{site.url}}{{site.baseurl}}/field-types/#get-a-mapping) operation, as shown in the following example request:
```

natebower commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,87 @@
+## Retrieving meta information
+
+You can retrieve the `_meta` information for an index using the [get mapping API]({{site.url}}{{site.baseurl}}/field-types/#get-a-mapping) operation, for example, as shown in the following request:
+
```
Line above: Confirm that "get mapping" shouldn't be capitalized.

natebower commented on _field-types/metadata-fields/meta.md:
````diff
@@ -0,0 +1,87 @@
+```
+{% include copy-curl.html %}
+
+This returns the full mapping for the index, including the `_meta` field, for example, as shown in the following response: 
````
```suggestion
The response returns the full index mapping, including the `_meta` field: 
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+# Routing
+
+OpenSearch uses a hashing algorithm to route documents to specific shards within an index. By default, the document's `_id` field is used as the routing value, but you can also specify a custom routing value for each document.
```
```suggestion
OpenSearch uses a hashing algorithm to route documents to specific shards in an index. By default, the document's `_id` field is used as the routing value, but you can also specify a custom routing value for each document.
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+## Default routing
+
+The following formula is the default routing formula used in OpenSearch. The `_routing` value is the document's `_id`. 
```
```suggestion
The following is the default OpenSearch routing formula. The `_routing` value is the document's `_id`. 
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+## Custom routing
+
+You can specify a custom routing value when indexing a document, as shown in the following example: 
```
```suggestion
You can specify a custom routing value when indexing a document, as shown in the following example request: 
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+In this example, the document is routed using the value `JohnDoe1` instead of the default `_id`.
+
+You must provide the same routing value when retrieving, deleting, or updating the document, as shown in the following example:
```
```suggestion
You must provide the same routing value when retrieving, deleting, or updating the document, as shown in the following example request:
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+## Querying by routing
+
+You can query documents based in their routing value using the `_routing` field, as shown in the following example. This query only searches the shard(s) associated with the `JohnDoe1` routing value.
```
```suggestion
You can query documents based on their routing value by using the `_routing` field, as shown in the following example. This query only searches the shard(s) associated with the `JohnDoe1` routing value:
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+## Required routing
+
+You can make custom routing a required field for all CRUD operations on an index, as shown in the following example. If you try to index a document without providing a routing value, OpenSearch throws an exception.
```
```suggestion
You can make custom routing a required field for all CRUD operations on an index, as shown in the following example request. If you try to index a document without providing a routing value, OpenSearch will throw an exception.
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+## Required routing
+
+You can make custom routing a required field for all CRUD operations on an index, as shown in the following example. If you try to index a document without providing a routing value, OpenSearch throws an exception.
+.
```
```suggestion
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+You can make custom routing a required field for all CRUD operations on an index, as shown in the following example. If you try to index a document without providing a routing value, OpenSearch throws an exception.
+.
+
```
Line 59: Should "custom routing" be in code font if it's a field name, or should it just read "You can make custom routing required for all CRUD operations on an index"?

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+## Routing to index partitions
+
+You can configure an index to route documents to a subset of shards, rather than a single shard. This is done using the `index.routing_partition_size` setting, as shown in the following example: 
```
```suggestion
You can configure an index to route documents to a subset of shards rather than a single shard by using the `index.routing_partition_size` setting, as shown in the following example request: 
```

natebower commented on _field-types/metadata-fields/routing.md:
````diff
@@ -0,0 +1,95 @@
+```
+{% include copy-curl.html %}
+
+## Routing to index partitions
````
Should this be something like "Routing to specific shards" to match the following text?

natebower commented on _field-types/metadata-fields/routing.md:
````diff
@@ -0,0 +1,95 @@
+```
+{% include copy-curl.html %}
+
+In this example, the documents are routed to one of four partitions within the index, based on a combination of the `_routing` and `_id` fields.
````
```suggestion
In this example, the documents are routed to one of four partitions in the index, based on a combination of the `_routing` and `_id` fields.
```

natebower commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,95 @@
+
+In this example, the documents are routed to one of four partitions within the index, based on a combination of the `_routing` and `_id` fields.
+
+Partitioned indexes have some limitations, such as not supporting `join` field relationships in the mappings.
```
_Or_ should line 76 talk about index partitions rather than shards? I'm most interested in phrasing consistency here.

vagimeli replied:
Revised to clarify. The term should be "routing partitions" instead of "partitions" in line 92.

natebower commented on _field-types/metadata-fields/source.md:
```diff
@@ -0,0 +1,54 @@
+
+## Including or excluding fields
+
+You can selectively control the contents of the `_source` field by using the `includes` and `excludes` parameters. This allows you to prune the stored `_source` field after indexing, but before it is saved. See the following example request:
```
```suggestion
You can selectively control the contents of the `_source` field by using the `includes` and `excludes` parameters. This allows you to prune the stored `_source` field after it is indexed but before it is saved, as shown in the following example request:
```

natebower commented on _field-types/mappings-use-cases.md:
```diff
@@ -0,0 +1,122 @@
+
+## Mapping string fields to `text` and `keyword` types
+
+To create an index named `movies1` with a dynamic template that maps all string fields to both `text` and `keyword` types, you can use the following request:
```
```suggestion
To create an index named `movies1` with a dynamic template that maps all string fields to both the `text` and `keyword` types, you can use the following request:
```

natebower commented on _field-types/metadata-fields/id.md:
````diff
@@ -0,0 +1,86 @@
+```
+{% include copy-curl.html %}
+
+You can then query the documents using the `_id` field, as shown in the following request:
````
```suggestion
You can then query the documents using the `_id` field, as shown in the following example request:
```

natebower commented on _field-types/metadata-fields/index-metadata.md:
````diff
@@ -0,0 +1,86 @@
+```
+{% include copy-curl.html %}
+
+You can then query both indexes and filter the results using the `_index` field, as shown in the following request:
````
```suggestion
You can then query both indexes and filter the results using the `_index` field, as shown in the following example request:
```

natebower commented on _field-types/metadata-fields/source.md:
```diff
@@ -0,0 +1,54 @@
+
+## Disabling the field
+
+You can disable the `_source` field by setting the `enabled` parameter to `false`, as shown in the following request:
```
```suggestion
You can disable the `_source` field by setting the `enabled` parameter to `false`, as shown in the following example request:
```

natebower commented on _field-types/metadata-fields/source.md:
````diff
@@ -0,0 +1,54 @@
+```
+{% include copy-curl.html %}
+
+Although these fields are not stored in the `_source`, you can still search on them, as the data is still indexed.
````
```suggestion
These fields are not stored in the `_source`, but you can still search them because the data remains indexed.
```

vagimeli commented on _field-types/index.md:
```diff
@@ -12,43 +12,77 @@ redirect_from:
+`double` | A double-precision, 64-bit floating-point number.
+`integer` | A signed 32-bit number.
+`object` | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.
+`array` | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements---not as fields with no values. OpenSear ...
```
```suggestion
`array` | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements---not as fields with no values. OpenSearch supports querying and filtering arrays, including checking for values, range queries, and array operations like concatenation and intersection. Nested arrays, which may contain complex objects or other arrays, can also be used for advanced data modeling. 
```

vagimeli commented on _field-types/index.md:
```diff
@@ -170,29 +151,31 @@ To get all mappings for one or more indexes, use the following request:
-Both `<index>` and `<fields>` can be specified as one value or a comma-separated list.
-
-For example, the following request retrieves the mapping for the `year` and `age` fields in `sample-index1`:
+Both `<index>` and `<fields>` can be specified as one value or a comma-separated list. For example, the following request retrieves the mapping for the `year` and `age` fields in `sample-index1`:
```
```suggestion
Both `<index>` and `<fields>` can be specified as either one value or a comma-separated list. For example, the following request retrieves the mapping for the `year` and `age` fields in `sample-index1`:
```

vagimeli commented on _field-types/metadata-fields/field-names.md:
```diff
@@ -0,0 +1,43 @@
+
+The `_field_names` field indexes field names that contain non-null values. This enables the use of the `exists` query, which can identify documents that either have or do not have non-null values for a specified field. 
+
+However, `_field_names` only indexes field names when both `doc_values` and `norms` are disabled for those fields. If either `doc_values` or `norms` are enabled, the `exists` query still functions but will not rely on the `_field_names` field.
```
```suggestion
However, `_field_names` only indexes field names when both `doc_values` and `norms` are disabled. If either `doc_values` or `norms` are enabled, then the `exists` query still functions but will not rely on the `_field_names` field.
```

vagimeli commented on _field-types/metadata-fields/id.md:
```diff
@@ -0,0 +1,86 @@
+
+# ID
+
+Each document in OpenSearch has a unique `_id` field. This field is indexed, allowing you to retrieve documents using the `GET` API or the [`ids` query]({{site.url}}{{site.baseurl}}/query-dsl/term/ids/).
```
```suggestion
Each document in OpenSearch has a unique `_id` field. This field is indexed, allowing you to retrieve documents using the GET API or the [`ids` query]({{site.url}}{{site.baseurl}}/query-dsl/term/ids/).
```

vagimeli commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,87 @@
+In this example, three custom metadata fields are added: application, version, and author. These fields can be used by your application to store any relevant information about the index, such as the application it belongs to, the version of the application, or the author of the index.
+
+You can update the `_meta` field using the [put mapping API]({{site.url}}{{site.baseurl}}/api-reference/index-apis/put-mapping/) operation, for example, as shown in the following request:
+
```
Followed style used in https://opensearch.org/docs/latest/api-reference/index-apis/put-mapping/

vagimeli commented on _field-types/metadata-fields/meta.md:
````diff
@@ -0,0 +1,87 @@
+```
+{% include copy-curl.html %}
+
+## Retrieving meta information
````
```suggestion
## Retrieving `meta` information
```

vagimeli commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,87 @@
+
+In this example, three custom metadata fields are added: `application`, `version`, and `author`. These fields can be used by your application to store any relevant information about the index, such as the application it belongs to, the application version, or the author of the index.
+
+You can update the `_meta` field using the [put mapping API]({{site.url}}{{site.baseurl}}/api-reference/index-apis/put-mapping/) operation, as shown in the following example request:
```
```suggestion
You can update the `_meta` field using the [Put Mapping API]({{site.url}}{{site.baseurl}}/api-reference/index-apis/put-mapping/) operation, as shown in the following example request:
```

vagimeli commented on _field-types/metadata-fields/meta.md:
```diff
@@ -0,0 +1,87 @@
+
+## Retrieving `meta` information
+
+You can retrieve the `_meta` information for an index using the [get mapping API]({{site.url}}{{site.baseurl}}/field-types/#get-a-mapping) operation, as shown in the following example request:
```
```suggestion
You can retrieve the `_meta` information for an index using the [Get Mapping API]({{site.url}}{{site.baseurl}}/field-types/#get-a-mapping) operation, as shown in the following example request:
```

vagimeli commented on _field-types/metadata-fields/routing.md:
````diff
@@ -0,0 +1,94 @@
+```
+{% include copy-curl.html %}
+
+## Routing to index partitions
````
```suggestion
## Routing to specific shards
```

vagimeli commented on _field-types/metadata-fields/routing.md:
````diff
@@ -0,0 +1,94 @@
+```
+{% include copy-curl.html %}
+
+In this example, the documents are routed to one of four partitions in the index, based on a combination of the `_routing` and `_id` fields.
````
```suggestion
In this example, the documents are routed to one of four routing partitions in the index, based on a combination of the `_routing` and `_id` fields.
```

vagimeli commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,94 @@
+
+In this example, the documents are routed to one of four routing partitions in the index, based on a combination of the `_routing` and `_id` fields.
+
+Partitioned indexes have some limitations, such as not supporting `join` field relationships in the mappings.
```
```suggestion
Index partitioning has limitations, such as not supporting `join` field relationships in the mappings.
```

vagimeli commented on _field-types/metadata-fields/routing.md:
````diff
@@ -0,0 +1,94 @@
+```
+{% include copy-curl.html %}
+
+In this example, the documents are routed to one of four routing partitions in the index, based on a combination of the `_routing` and `_id` fields.
````
```suggestion
In this example, the documents are routed to one of four shards in the index, based on a combination of the `_routing` and `_id` fields.
```

vagimeli commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,94 @@
+
+In this example, the documents are routed to one of four shards in the index, based on a combination of the `_routing` and `_id` fields.
+
+Index partitioning has limitations, such as not supporting `join` field relationships in the mappings.
```
```suggestion
```

vagimeli commented on _field-types/metadata-fields/routing.md:
```diff
@@ -0,0 +1,92 @@
+
+## Routing to specific shards
+
+You can configure an index to route custom values to a subset of shards rather than a single shard. This is done by setting `index.routing_partition_size` at the time of index creation. With this setting, the formula for calculating the shard is `shard_num = (hash(_routing) + hash(_id)) % routing_partition_size) % num_primary_shards`.
```
```suggestion
You can configure an index to route custom values to a subset of shards rather than a single shard. This is done by setting `index.routing_partition_size` at the time of index creation. The formula for calculating the shard is `shard_num = (hash(_routing) + hash(_id)) % routing_partition_size) % num_primary_shards`.
```
